Conan Release Builder (2.4.0 sync)

tequdev opened the pull request:
<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.
-->

## High Level Overview of Change

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

<!--
Please check [x] relevant options, delete irrelevant ones.

* If there is any impact to the public API methods (HTTP / WebSocket), please update https://github.com/xrplf/rippled/blob/develop/API-CHANGELOG.md
  * Update API-CHANGELOG.md and add the change directly in this PR by pushing to your PR branch.
* libxrpl: See https://github.com/XRPLF/rippled/blob/develop/docs/build/depend.md
* Peer Protocol: See https://xrpl.org/peer-protocol.html
-->

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

tequdev committed:
use conan for release-builer

tequdev committed:
Merge remote-tracking branch 'upstream/sync-2.4.0' into sync-2.4.0-conan-release-builder

tequdev committed:
fix

tequdev committed:
include tests

tequdev committed:
ninja

tequdev committed:
Image size reduction

tequdev committed:
Merge branch 'sync-2.4.0' into sync-2.4.0-conan-release-builder

tequdev committed:
Merge branch 'sync-2.4.0' into sync-2.4.0-conan-release-builder

tequdev committed:
Merge branch 'sync-2.4.0' into sync-2.4.0-conan-release-builder

sublimator commented on release-builder.sh:
```diff
@@ -1,9 +1,8 @@
-#!/bin/bash -u
+#!/bin/bash 
```
Unsure why removed the -u and set -ex yet as haven't read down, but the comment remains

tequdev replied:
I deleted it because an error occurred when executing locally because some GitHub ENV did not exist in local

sublimator commented on release-builder.sh:
```diff
@@ -50,25 +49,187 @@ STATIC_CONTAINER=$(docker ps -a | grep $CONTAINER_NAME |wc -l)
+
+# Configure ccache and Conan
+RUN /hbb_exe/activate-exec bash -c "ccache -M 10G && \
+    ccache -o cache_dir=/cache/ccache && \
```
Note that unless you configure ccache to use content tracking, afaik it's not very useful in a CI situation due to git checkouts updating the modification times of files


> By default, ccache uses mtimes (modification times) to detect changes. Git checkouts reset timestamps, making ccache think everything changed even if content is identical. Setting compiler_check = content makes ccache hash the actual file contents instead.

tequdev replied:
I tried it locally, but I couldn't see any difference in behavior between the default and compiler_check=content.

sublimator replied:
I mean, it's a moot point when there's nothing cached
It's only when you're able to reuse the container somehow

tequdev replied:
https://github.com/Xahau/xahaud/pull/516/commits/ae1c906e0348664751b852ed1242b5468f7cf089

sublimator commented on release-builder.sh:
```diff
@@ -50,25 +49,187 @@ STATIC_CONTAINER=$(docker ps -a | grep $CONTAINER_NAME |wc -l)
+# Build and install WasmEdge
+RUN cd /tmp && \
+    ( wget -nc -q https://github.com/WasmEdge/WasmEdge/archive/refs/tags/0.11.2.zip; unzip -o 0.11.2.zip; ) && \
+    cd WasmEdge-0.11.2 && \
```
When you install this and boost like so will conan install --missing will skip these?

sublimator replied:
Also curious why using these over the conan versions ? Specific build flags?

sublimator replied:
'-D_GLIBCXX_USE_CXX11_ABI=0'

But then if Boost/WasmEdge need old ABI, don't ALL the C++ dependencies need it too?

tequdev replied:
conan install will not skip those.
However, wasmedge will use what was built here in release-build. boost should be used the one from conan.

To use it as a static lib, building wasmedge here.
The one provided by conan is a shared lib.

sublimator replied:
> The one provided by conan is a shared lib.

I wonder if that can be configured in the conan file somehow

sublimator replied:
> boost should be used the one from conan.

Ideally there's only one version of boost linked, but I confess I'm not an expert on linking so I don't know exactly what could go wrong there

sublimator replied:
Given it's static, it's likely not a problem, is it? Because the final linking isn't done until an actual binary is made (rippled), where it's going to use the conan boost. Maybe some template specialization issues.

tequdev replied:
https://github.com/Xahau/xahaud/pull/516/commits/01a3c4c45160fbbb7d5ebca954d4a0db08a2a71c

sublimator commented on release-builder.sh:
```diff
@@ -50,25 +49,187 @@ STATIC_CONTAINER=$(docker ps -a | grep $CONTAINER_NAME |wc -l)
+ENV Boost_LIBRARY_DIRS=/usr/local/lib
+ENV BOOST_INCLUDEDIR=/usr/local/src/boost_1_86_0
+
+ENV CMAKE_CXX_FLAGS='-D_GLIBCXX_USE_CXX11_ABI=1'
```
New ABI here though?
cmake will honor "CMAKE_CXX_FLAGS" as if its CXX_FLAGS and pass along?

tequdev replied:
removed

tequdev committed:
Merge branch 'sync-2.4.0' into sync-2.4.0-conan-release-builder

tequdev committed:
update to based on tequdev:conan-release-builder

tequdev committed:
use mold linker for release-builder

- Optimize ccache
- Processing time reduced from approximately 3 mins and 40 secs to approximately 1 min and 20 secs when running in a local machine.

tequdev committed:
Make release build use conan deps where possible and hbb 4.0.1 (#516)


Co-authored-by: Denis Angell <dangell@transia.co>
Co-authored-by: Niq Dudfield <ndudfield@gmail.com>

tequdev committed:
Merge remote-tracking branch 'upstream/dev' into sync-2.4.0-conan-release-builder
